Refactor emit snapshot

pgrange opened the pull request:
[Inline emit snapshot handling on AckSn and ReqTx handling](https://github.com/input-output-hk/hydra/pull/980/commits/8478510c5bb3b98765f7c44368ec169167cf75be) 
- Removed tautology (T ̸ ˆ = ∅) introduced to make the emit snapshot reusable during ReqTx handling
- Removed tautology (sˆ = ¯s) introduced to make the emit snapshot reusable during AckSn handling


With this in place, now we do not need to compute the next head state to decide wheter or not to emit a snapshot.
This enables to refactor, more easily, the head logic towards event-sourcing (ADR#24)

---

* [x] CHANGELOG updated or not needed
* [x] Documentation updated or not needed
* [x] Haddocks updated or not needed
* [x] No new TODOs introduced or explained herafter

ffakenz committed:
Refactor tests to use update instead of emitSnapshot

ffakenz committed:
Inline emit snapshot handling on AckSn and ReqTx handling

- Removed tautology (sˆ = ¯s) introduced to make the emit snapshot reusable during AckSn handling
- Removed tautology (T ̸ ˆ = ∅) introduced to make the emit snapshot reusable during ReqTx handling

With this in place, now we do not need to compute the next head state to decide wheter or not to emit a snapshot.
This enables to refactor, more easily, the head logic towards event-sourcing (ADR#24)

pgrange committed:
Update the spec

pgrange commented on hydra-node/src/Hydra/HeadLogic.hs:
```diff
@@ -426,6 +426,16 @@ collectWaits = \case
   Effects _ -> []
   Combined l r -> collectWaits l <> collectWaits r
 
+-- FIXME should return Maybe?
```
Should fix this first.

ffakenz replied:
we decided that we are happy as it is.

pgrange replied:
```suggestion
```

pgrange commented on hydra-node/src/Hydra/HeadLogic.hs:
```diff
@@ -646,35 +656,67 @@ onOpenNetworkReqTx env ledger st ttl tx =
-                    { seenTxs = seenTxs <> [tx]
-                    , seenUTxO = utxo'
+      Effects [ClientEffect $ TxValid headId tx]
+        `Combined` if isLeader parameters party nextSn && not snapshotInFlight
```
```suggestion
        `Combined` if isLeader parameters party nextSn && confirmedSn == seenSn
```

We should define seenSn in terms of seenSnapshot

pgrange replied:
On master, the `seenSnapshot` function is

```
-- | Get the last seen snapshot number given a 'SeenSnapshot'.
seenSnapshotNumber :: SeenSnapshot tx -> SnapshotNumber
seenSnapshotNumber = \case
  NoSeenSnapshot -> 0
  LastSeenSnapshot{lastSeen} -> lastSeen
  RequestedSnapshot{lastSeen} -> lastSeen
  SeenSnapshot{snapshot = Snapshot{number}} -> number
```

For this change proposal to work, we would need to change this function to be:

```
-- | Get the last seen snapshot number given a 'SeenSnapshot'.
seenSnapshotNumber :: SeenSnapshot tx -> SnapshotNumber
seenSnapshotNumber = \case
  NoSeenSnapshot -> 0
  LastSeenSnapshot{lastSeen} -> lastSeen
  RequestedSnapshot{requested} -> requested
  SeenSnapshot{snapshot = Snapshot{number}} -> number
```

But this breaks the code because we also use this function in `onOpenNetworkReqSn` where we really want the master's way of computing the seenSnapshot number. For instance, we can see that the following test becomes red:

```
  test/Hydra/NodeSpec.hs:69:7: 
  1) Hydra.Node emits a single ReqSn and AckSn as leader, even after multiple ReqTxs
       expected: [ReqSn {snapshotNumber = UnsafeSnapshotNumber 1, transactionIds = [1]},AckSn {signed = HydraSignature "4d5e417b001974f93c8932ec29efb0c93c75de211c5413d78b806e6dbc41606d10c1c64735bb384b8b3ff2165ed46dbba9d9593a80f102285a8d9892e3ffe609", snapshotNumber = UnsafeSnapshotNumber 1}]
        but got: [ReqSn {snapshotNumber = UnsafeSnapshotNumber 1, transactionIds = [1]}]
```

Since this change is not part of this P.R. original scope, I say we don't do it here and create a red bin out of this comment to figure out what's going on with this seen snapshot concept here.

pgrange replied:
See #982

pgrange commented on hydra-node/src/Hydra/HeadLogic.hs:
```diff
@@ -646,35 +656,67 @@ onOpenNetworkReqTx env ledger st ttl tx =
+                          , seenUTxO = utxo'
+                          , seenSnapshot =
+                              RequestedSnapshot
+                                { lastSeen = seenSnapshotNumber seenSnapshot
```
We could use seenSn here if available (see above comment)
```suggestion
                                { lastSeen = seenSn
```

pgrange replied:
See above comment.

pgrange commented on hydra-node/src/Hydra/HeadLogic.hs:
```diff
@@ -823,21 +865,35 @@ onOpenNetworkAckSn env openState otherParty snapshotSignature sn =
+                        coordinatedHeadState
+                          { seenSnapshot =
+                              RequestedSnapshot
+                                { lastSeen = seenSnapshotNumber $ LastSeenSnapshot (number snapshot)
```
```suggestion
                                { lastSeen = sn
```

pgrange commented on hydra-node/src/Hydra/HeadLogic.hs:
```diff
@@ -823,21 +865,35 @@ onOpenNetworkAckSn env openState otherParty snapshotSignature sn =
+                                { snapshot
+                                , signatures = multisig
+                                }
+                          , seenSnapshot = LastSeenSnapshot (number snapshot)
```
```suggestion
                          , seenSnapshot = LastSeenSnapshot sn
```

pgrange commented on hydra-node/test/Hydra/HeadLogicSnapshotSpec.hs:
```diff
@@ -0,0 +1,244 @@
+      prop "there's always a leader for every snapshot number" prop_thereIsAlwaysALeader
+
+    describe "On ReqTx" $ do
+      prop "always emit ReqSn given head has 1 member" prop_singleMemberHeadAlwaysSnapshotOnReqTx
```
This property also holds for AckSn. We should try to adapt it

v0d1ch replied:
We agreed this can be also done separatelly.

pgrange commented on hydra-node/test/Hydra/HeadLogicSnapshotSpec.hs:
```diff
@@ -0,0 +1,244 @@
+        collectEffects outcome
+          `shouldContain` [NetworkEffect (ReqSn 1 [txId tx])]
+
+      it "Do NOT send ReqSn when we are NOT the leader even if no snapshot in flight" $ do
```
```suggestion
      it "do NOT send ReqSn when we are NOT the leader even if no snapshot in flight" $ do
```

pgrange commented on hydra-node/test/Hydra/HeadLogicSnapshotSpec.hs:
```diff
@@ -0,0 +1,244 @@
+
+        collectEffects outcome `shouldNotSatisfy` sendReqSn
+
+      it "Do NOT send ReqSn when we are the leader but snapshot in flight" $ do
```
```suggestion
      it "do NOT send ReqSn when we are the leader but snapshot in flight" $ do
```

pgrange commented on hydra-node/test/Hydra/HeadLogicSnapshotSpec.hs:
```diff
@@ -0,0 +1,244 @@
+        let outcome = update bobEnv simpleLedger headState $ ackFrom bobSk bob
+        collectEffects outcome `shouldNotSatisfy` sendReqSn
+
+      it "Do NOT send ReqSn when we are NOT the leader but there are seen transactions" $ do
```
```suggestion
      it "do NOT send ReqSn when we are NOT the leader but there are seen transactions" $ do
```

pgrange commented on hydra-node/test/Hydra/HeadLogicSpec.hs:
```diff
@@ -634,14 +635,9 @@ assertNewState outcome =
-    _ -> Nothing
+  case collectState outcome of
+    [newState] -> pure newState
+    _ -> Hydra.Test.Prelude.error $ "Expecting one single newState in outcome: " <> show outcome
```
Call failure here

pgrange reviewed (commented):
Notes when reviewing together

pgrange committed:
We're happy with it that way for now

This will be changed with the event-sourced approach anyway.

pgrange committed:
Pick the snapshot number right from the function's parameter

pgrange committed:
Fix grammar in tests description

ch1bo committed:
Fix state update on confirming snapshot

This was not covered by a test and hence we have created a bit more
higher level a behavior spec which asserts that snapshots are created
while there are still transactions seen and not in snapshots yet.

ch1bo committed:
Drop the pending test in BehaviorSpec

This is now covered by the recently added test about snapshotting there.

pgrange committed:
FIX performance issue

Forgot to purge the allTxs list when transaction is included in a
snapshot.

Added a test to ensure that we check that both, when we're supposed
to send a ReqSn and when we're not.

This is not pretty as we duplicate one test so that we can check
for every branch but it's safer now that we're explicit about this
and we can, hopefully, improve that when we migrate to event-sourced
approach.
